pkg/host-disk/host-disk.go

```diff
@@ -85,19 +92,58 @@ func getPVCDiskImgPath(volumeName string) string {
 	return path.Join(pvcBaseDir, volumeName, "disk.img")
 }
 
-func CreateHostDisks(vmi *v1.VirtualMachineInstance) error {
+type DiskImgCreator struct {
+	dirBytesAvailableFunc  func(path string) (uint64, error)
+	notifier               k8sNotifier
+	lessPvcSpaceToleration int
+}
+
+type k8sNotifier interface {
+	SendK8sEvent(vmi *v1.VirtualMachineInstance, severity string, reason string, message string) error
+}
+
+func NewHostDiskCreator(notifier k8sNotifier, lessPvcSpaceToleration int) DiskImgCreator {
+	return DiskImgCreator{
+		dirBytesAvailableFunc:  dirBytesAvailable,
+		notifier:               notifier,
+		lessPvcSpaceToleration: lessPvcSpaceToleration,
+	}
+}
+
+func (hdc *DiskImgCreator) setLessPvcSpaceToleration(toleration int) {
+	hdc.lessPvcSpaceToleration = toleration
+}
+
+func (hdc DiskImgCreator) Create(vmi *v1.VirtualMachineInstance) error {
 	for _, volume := range vmi.Spec.Volumes {
 		if hostDisk := volume.VolumeSource.HostDisk; hostDisk != nil && hostDisk.Type == v1.HostDiskExistsOrCreate && hostDisk.Path != "" {
 			if _, err := os.Stat(hostDisk.Path); os.IsNotExist(err) {
-				availableSpace, err := dirBytesAvailable(path.Dir(hostDisk.Path))
+				availableSize, err := hdc.dirBytesAvailableFunc(path.Dir(hostDisk.Path))
 				if err != nil {
 					return err
 				}
-				size, _ := hostDisk.Capacity.AsInt64()
-				if uint64(size) > availableSpace {
-					return fmt.Errorf("Unable to create %s, not enough space, demanded size %d B is bigger than available space %d B", hostDisk.Path, uint64(size), availableSpace)
+				requestedSize, _ := hostDisk.Capacity.AsInt64()
+				diskSize := requestedSize
+				if uint64(requestedSize) > availableSize {
+					// Some storage provisioners provision less space than requested, due to filesystem overhead etc.
+					// We tolerate some difference in requested and available capacity up to some degree.
+					// This can be configured with the "pvc-tolerate-less-space-up-to-percent" parameter in the kubevirt-config ConfigMap.
+					// It is provided as argument to virt-launcher.
+					toleratedSize := requestedSize * (100 - int64(hdc.lessPvcSpaceToleration)) / 100
+					if uint64(toleratedSize) > availableSize {
+						return fmt.Errorf("unable to create %s, not enough space, demanded size %d B is bigger than available space %d B, also after taking %v %% toleration into account",
+							hostDisk.Path, uint64(requestedSize), availableSize, hdc.lessPvcSpaceToleration)
+					}
+					diskSize = int64(availableSize)
+
+					msg := fmt.Sprintf("PV size too small: expected %v B, found %v B. Using it anyway, it is within %v %% toleration", requestedSize, availableSize, hdc.lessPvcSpaceToleration)
+					err = hdc.notifier.SendK8sEvent(vmi, EventTypeToleratedSmallPV, EventReasonToleratedSmallPV, msg)
+					if err != nil {
+						log.Log.Reason(err).Warningf("Couldn't send k8s event for tolerated PV size: %v", err)
+						log.Log.Info(msg)
```

Comment:
I think we want to log the message always.

Reply:
done